### PR TITLE
Add ruleset-level legacy skin provider

### LINKED ISSUE #5786: Make LegacySkin implementations per-ruleset
Currently it is not able to DI a DrawableHitObject and cast to a ruleset specific type, as `LegacySkin` is in the `osu.Game` project.

The proposal would be to have a method of making per-ruleset legacy skin implementations to work around this.

### PR DESCRIPTION
**Pull request by peppy**
Closes #5786.

This allows a ruleset to provide a custom implementation for legacy skinning.

The hierarchy is a bit convoluted from the change (due to the source being the parent `SkinProvidingContainer`) but this is the best way I can find for this for now to ensure correct lookups.

### CODE REVIEW

File: osu.Game/Skinning/BeatmapSkinProvidingContainer.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    /// <summary>
+    /// A container which overrides existing skin options with beatmap-local values.
+    /// </summary>
+    public class BeatmapSkinProvidingContainer : SkinProvidingContainer
+    {
+        private readonly Bindable<bool> beatmapSkins = new Bindable<bool>();
+        private readonly Bindable<bool> beatmapHitsounds = new Bindable<bool>();
+
+        protected override bool AllowConfigurationLookup => beatmapSkins.Value;
+        protected override bool AllowDrawableLookup(string componentName) => beatmapSkins.Value;
```
Comment (smoogipoo) [inline osu.Game/Skinning/BeatmapSkinProvidingContainer.cs]: Seems to break beatmap skins on first load (if you uncheck-check the checkbox it'll work). Probably due to load() order of the hierarchy.

**Review by smoogipoo — CHANGES_REQUESTED** [review]
Looks ok otherwise

File: osu.Game/Skinning/BeatmapSkinProvidingContainer.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Skinning/BeatmapSkinProvidingContainer.cs]: Fix via loading `DrawableRuleset` in a second pass.

**Comment by peppy** [comment]
Hold off on merging this one briefly. Going to look into how easy it is to add a player-based test against beatmap/skin providers.

**Comment by peppy** [comment]
Tests added.

File: osu.Game.Rulesets.Osu.Tests/TestSceneOsuPlayer.cs [hunk 2]
```diff
@@ -1,17 +1,157 @@
...
+            {
+                this.skin = skin;
+            }
+
+            protected override ISkin GetSkin() => skin;
+        }
+
+        public class SkinProvidingPlayer : TestPlayer
+        {
+            private readonly TestSource userSkin;
```
Comment (iiSaLMaN) [inline osu.Game.Rulesets.Osu.Tests/TestSceneOsuPlayer.cs]:
```suggestion
            [Cached(typeof(ISkinSource))]
            private readonly TestSource userSkin;
```

**Review by smoogipoo — APPROVED** [review]
Lgtm

**Pull request merged**